src/packages/core/src/metadata-service/resolver.utils.ts

```diff
@@ -0,0 +1,218 @@
+) => {
+	const {
+		resolverOptions: {
+			source,
+			args: { filter },
+		},
+		idValue,
+		relatedPrimaryKeyField,
+		relatedField,
+		relatedEntityMetadata,
+		sourcePrimaryKeyField,
+	} = params;
+	const _and: Filter<R>[] = [];
+
+	// If we have a user supplied filter, add it to the _and array.
+	if (filter) _and.push(filter);
+
+	// Lets check the relationship type and add the appropriate filter.
+	let relationshipFilterChunk: Filter<R> | undefined = undefined;
+
+	if (idValue) {
+		if (Array.isArray(idValue)) {
+			relationshipFilterChunk = { [`${relatedPrimaryKeyField}_in`]: idValue } as Filter<R>;
+		} else {
+			relationshipFilterChunk = { [relatedPrimaryKeyField]: idValue } as Filter<R>;
+		}
+	} else if (
+		relatedField &&
+		isScalarType(getFieldType(relatedEntityMetadata.fields[relatedField]))
+	) {
+		// Scalars should get a simple filter, e.g. if we have Tasks in a DB which have a userId field, and we
+		// have Users in a REST API with their PK called 'key', instead of trying to filter like:
+		// { userId: { key: '1' } }
+		// we should instead filter like:
+		// { userId: '1' }
+		// because the database provider doesn't understand the shape of the user object at all.
+		relationshipFilterChunk = {
+			[relatedField]: source[sourcePrimaryKeyField],
+		} as unknown as Filter<R>;
+	} else if (relatedField) {
+		// While object filters should nest as not all providers understand what { user: '1' } means. It's more
+		// clear to give them { user: { key: '1' } }.
+		relationshipFilterChunk = {
+			[relatedField]: { [sourcePrimaryKeyField]: source[sourcePrimaryKeyField] },
+		} as Filter<R>;
+	} else {
+		throw new Error(
+			'Did not determine how to filter the relationship. Either id or relatedField is required.'
+		);
+	}
+
+	if (relationshipFilterChunk) _and.push(relationshipFilterChunk);
+
+	const relatedEntityFilter = { _and } as Filter<R>;
+
+	return { relatedEntityFilter, relationshipFilterChunk };
+};
+
+/**
+ * To load the entities in bulk, we need a filter that doesn't contain the relationship filter.
+ * This function will remove the relationship filter from the filter.
+ */
+export const getLoaderFilter = <R>(
+	hookParams: ReadHookParams<R, object>,
+	relationshipFilterChunk: Filter<R>
+) => {
+	let relationshipFilterChunkFound = !relationshipFilterChunk;
+	/**
+	 * Look for `relationshipFilterChunk` and remove it from the filter.
+	 * This is necessary because we are batch loading data and that relationshipFilterChunk makes the filter different for each record, which ends up triggering a SQL query for each record (not batching at all).
+	 * What we want to achieve is having a filter with ACLs into it, but without the relationship chunk.
+	 */
+	const removeRelationshipFilterChunk = (
+		filter: Filter<R> | undefined,
+		relationshipFilterChunk: Filter<R>
+	): Filter<R> | undefined => {
+		if (!filter?._and) return filter;
+		return {
+			_and: filter._and.map((item) => {
```

Comment:
Probably worth adding a comment here that the reason we're only looking through `_and` is because the core code puts the join in an `_and` filter. Other filters are supported and will work fine with this code, we just know that the relationship filter won't be in there.